### ISSUE TITLE
Massive Speedup

### ISSUE THREAD

**Issue by TryBane** [29988122/Fate-Grand-Order_Lua#113]
I've been fiddling with the main script last night and managed to make some tweaks that reliably allowed for faster runs on my phone.

This fiddling relies less on wait() and moves to checking for when a skill is complete by searching for the same image as we check for at the start of each stage. This allows slower phones to still reliably use the script as is, and speeds up the script for faster processing speeds. I will post the fiddled with regular script if anyone would be willing to give it a test to ensure it still works reliably on more than my own device.

An important note, this only speeds up the activation of skills, clicking of NPs, and makes the result screen script more reliable (I was getting interrupts in the instance of leveling my Mystic Code) 

--EDIT Also need to check if it works reliably with slow battle speed, since I used it with fast.

**Comment by potchy** [comment]
I think we don't need to even consider slow battle speed.
There's literally no need to use it when "botting".

**Comment by 29988122** [comment]
A better way would be that we normalize(abstract?) the wait() in our script, so changes can be made easily instead of hard-coded in our routine.

After that we can measure the length of skill animation, card clicking speed, or even average battle animation, to name a few.

Although I'd personally put it on the bottom of the priority list, XD
(Won't have enough free time in the foreseeable future - at least for 3 months. And I guess Potchy & ryugawa has their own module to maintain.)

**Comment by ParasiteEve** [comment]
@TryBane how did you speed up the skills? I tried looking for the correct wait() modifier but was unable to, thank you

**Comment by TryBane** [comment]
@ParasiteEve I noticed that every time a skill is cast, the battle.png image that is used to determine that a battle has begun in the script, disappears from the screen during the animation. What I did was set a small period to wait ( .4 seconds if I recall correctly) then keep checking if the image is back within a loop. After it is found, it waits another .4 seconds before activating the next skill.

I did this without looking at the API though, just going off of my own programming experience and the syntax I noticed being used within the Lua script as of now. However, I found a few options that are likely better to use. And this is why I posted a new topic #114 because I feel that the entire script can be modified to be much faster, efficient, flexible, and less error prone.

If you are still curious, the option I found I will probably use is the "waitVanish()" to wait until the image disappears from site, then the "wait()" as normal, except it will be searching for the image itself rather than for a specified amount of time. After the image is found, since there is a slight delay between activating a skill when it appears, I would set an additional wait() for .2-.4 seconds. These two functions will be processed back to back to ensure that a skills animation completed before moving on to the next skill. It is important to note that when you choose a targetable skill, the battle.png image does NOT disappear when the menu to choose a servant to activate the skill on comes up. So I plan to implement another image search looking for the x button in the top right corner to determine if that menu is up, and if it is, to then select the servant to use the skill on.

For anyone who has worked with the script, you know that this isn't easy to accomplish with how the script is setup now. Which is why I want to do an overhaul.
